### PR TITLE
References to old and new blocks were mixed in xlog_heap_update handler

### PR DESCRIPTION
**Pull request by knizhnik**
## Problem

See https://neondb.slack.com/archives/C05L7D1JAUS/p1694614585955029
https://www.notion.so/neondatabase/Duplicate-key-issue-651627ce843c45188fbdcb2d30fd2178

## Summary of changes

Swap old/new block references

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Review by save-buffer — APPROVED** [review]
Seems simple enough

**Comment by hlinnaka** [comment]
I added a python test here. However, at least on my laptop it's still failing after with this fix. I didn't investigate why; could be a bug in my test..

File: pageserver/src/walingest.rs [hunk 1]
```diff
@@ -470,14 +470,14 @@ impl<'a> WalIngest<'a> {
...
-                            old_heap_blkno = Some(decoded.blocks[0].blkno);
+                            old_heap_blkno = Some(decoded.blocks[1].blkno);
                         }
                         if (xlrec.flags & pg_constants::XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED) != 0 {
                             // PostgreSQL only uses XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED on a
                             // non-HOT update where the new tuple goes to different page than
                             // the old one. Otherwise, only XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED is
                             // set.
-                            new_heap_blkno = Some(decoded.blocks[1].blkno);
+                            new_heap_blkno = Some(decoded.blocks[0].blkno);
```
Comment (hlinnaka) [inline pageserver/src/walingest.rs:480]: Hmm, what happens on a same-page update now? Reading the code in heap_update(), I think if the old and new tuple are on same page:
- only block ref 0 is used
- XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED is set.

Isn't the `old_heap_blkno = Some(decoded.blocks[1].blkno);` assignment wrong in that case?

I'm surprised none of the existing tests catch that.

Comment (hlinnaka) [reply pageserver/src/walingest.rs:480]: Expanded the test case to cover that case. Indeed, you now get an error:
```
2023-09-14T21:57:56.770817Z ERROR wal_connection_manager{tenant_id=34c2e198bfd42a89284d97c4aa26aeba timeline_id=613e7b5e2bab2a021d30addf2fea084f}:connection{node_id=1}:panic{thread=walreceiver worker location=pageserver/src/walingest.rs:529:51}: index out of bounds: the len is 1 but the index is 1
```

**Comment by hlinnaka** [comment]
> I added a python test here. However, at least on my laptop it's still failing after with this fix. I didn't yet investigate why; could be a bug in my test..

That's because the v15 variant of the code hasn't been fixed yet. It's annoying that we have multiple copy-pasted variants of the same code. A refactoring would be in order, but let's fix this critical bug first.

**Review by hlinnaka — APPROVED** [review]

**Pull request merged**